src/ray/pubsub/publisher.cc

```diff
@@ -20,149 +20,188 @@ namespace pubsub {
+  auto &subscribers = message_id_to_subscriber_it->second;
   auto subscriber_it = subscribers.find(subscriber_id);
   // If code reaches this line, that means the subscriber id was in the index.
   RAY_CHECK(subscriber_it != subscribers.end());
   subscribers.erase(subscriber_it);
   if (subscribers.size() == 0) {
-    objects_to_subscribers_.erase(objects_to_subscribers_it);
+    message_id_to_subscribers_.erase(message_id_to_subscriber_it);
   }
   return true;
 }
 
-bool SubscriptionIndex::AssertNoLeak() const {
-  return objects_to_subscribers_.size() == 0 && subscribers_to_objects_.size() == 0;
+template <typename MessageID>
+bool SubscriptionIndex<MessageID>::CheckNoLeaks() const {
+  return message_id_to_subscribers_.size() == 0 && subscribers_to_message_id_.size() == 0;
 }
 
-bool Subscriber::ConnectToSubscriber(
-    LongPollConnectCallback long_polling_reply_callback) {
-  if (long_polling_reply_callback_ == nullptr) {
-    long_polling_reply_callback_ = long_polling_reply_callback;
+bool Subscriber::ConnectToSubscriber(rpc::PubsubLongPollingReply *reply,
+                                     rpc::SendReplyCallback send_reply_callback) {
+  if (!long_polling_connection_) {
+    RAY_CHECK(reply != nullptr);
+    RAY_CHECK(send_reply_callback != nullptr);
+    long_polling_connection_ =
+        absl::make_unique<LongPollConnection>(reply, std::move(send_reply_callback));
     last_connection_update_time_ms_ = get_time_ms_();
     return true;
   }
   return false;
 }
 
-void Subscriber::QueueMessage(const ObjectID &object_id, bool try_publish) {
-  mailbox_.push_back(object_id);
+void Subscriber::QueueMessage(std::unique_ptr<rpc::PubMessage> pub_message,
+                              bool try_publish) {
+  if (mailbox_.size() == 0) {
+    mailbox_.push_back(absl::make_unique<rpc::PubsubLongPollingReply>());
+  }
+
+  // Update the long polling reply.
+  auto *next_long_polling_reply = mailbox_.back().get();
+  auto *new_pub_message = next_long_polling_reply->add_pub_messages();
+  new_pub_message->Swap(pub_message.get());
+
+  // If the reply is too big, add a new entry.
+  if (next_long_polling_reply->pub_messages_size() >= publish_batch_size_) {
+    mailbox_.push_back(absl::make_unique<rpc::PubsubLongPollingReply>());
+  }
+
   if (try_publish) {
     PublishIfPossible();
   }
 }
 
 bool Subscriber::PublishIfPossible(bool force) {
-  if (long_polling_reply_callback_ == nullptr) {
+  if (!long_polling_connection_) {
     return false;
   }
 
   if (force || mailbox_.size() > 0) {
-    std::vector<ObjectID> mails_to_post;
-    while (mailbox_.size() > 0 && mails_to_post.size() < publish_batch_size_) {
-      // It ensures that mails are posted in the FIFO order.
-      mails_to_post.push_back(mailbox_.front());
-      mailbox_.pop_front();
+    // If force pubilsh is invoked, mailbox could be empty. Fill this up in that case.
+    if (mailbox_.size() == 0) {
+      mailbox_.push_back(absl::make_unique<rpc::PubsubLongPollingReply>());
```

Comment:
This part is wired here. Should we remove this? If it's empty, should we just skip sending the message?